### ISSUE TITLE
Select columns to write in write.fst

### ISSUE THREAD

**Issue by mgahan** [fstpackage/fst#124]
Would it be possible to select what columns to write in `write.fst` similar to how `read.fst` allows for the reading of specific columns? I know this probably isn't a common request but it would be useful in my workflow (but not if is would add too much overhead or time cost).

**Comment by MarcusKlik** [comment]
Hi @mgahan, thanks for your request!

There are several routes you can take to select (or rename) columns and rows before writing your table to disk:

``` r

library(data.table)
library(dplyr)
library(fst)

# sample data.table
dt <- data.table(X = 1:26, Y = LETTERS)

# directly using data.table's interface
write_fst(dt[, .(X)], "1.fst")

# rename column and use pipe operator
dt[, .(Z = X)] %>%
  write_fst("1.fst")

# add a row selection
dt[X < 5, .(Z = X)] %>%
  write_fst("1.fst")

# do the same with a tibble
tbl <- tibble(X = 1:26, Y = LETTERS)

tbl %>%
  filter(X < 5) %>%
  select(Z = X) %>%
  write_fst("2.fst")
```

Because the interfaces of `data.table` and `dplyr` are so rich with options, I think it will be hard for the `fst` package to be as flexible. A column selection would be possible in the `fst` interface, but then you would miss out on renaming or row selection, what do you think?

**Comment by shrektan** [comment]
I agree. We’d better make the interface of fst as simple as possible. Considering existing so many alternative ways of doing selecting columns, I prefer not make things complicated.

**Comment by mgahan** [comment]
My main goal is to avoid making copies of any data if at all possible (since the data is very big and any copies are pretty wasteful). Do you know if any of those methods avoid copying the data? Anyways, I understand if it isn’t a feature most people would want. I understand the idea for it to be simple. Thanks so much for the great package!

**Comment by renkun-ken** [comment]
`DT[, .(x, y, z)]` indeed makes a copy of each vector. Imagine we need to write 300 columns of a 301-column `DT` of many many rows.

A simple subsetting method without triggering copy is as follows:

```r
dt <- .subset(DT, c("x", "y", "z"))
setDT(dt)
```

which creates a shallow copy of `DT` with only selected columns but `DT`'s key is unfortunately lost. And `dt` must be used with cautious since `dt$x` is exactly `DT$x` and the same for `y` and `z`. Any in-place reordering of `dt` (e.g. `setkey(dt, y, x)`) would cause data mess up in `DT`.

**Comment by xiaodaigh** [comment]
If there is no way to avoid the copying then I would vote for a feature in fst that writes only certain column. Accessing a column is "free" as it's just a pointer. Is there not a way to create a new dataframe by just providing it with pointers to the columns? I would make sure that that doesn't exist first. I think it exists because data.table can create a new column in a data.frame without creating a new data.frame

**Comment by MarcusKlik** [comment]
Hi @xiaodaigh, @renkun-ken and @mgahan, indeed, there is a clear use-case then!

Just to check:

``` r
library(data.table)

dt <- data.table(X = 1:26, Y = LETTERS)

tracemem(dt)
#> [1] "<0000000017509998>"
tracemem(dt$X)
#> [1] "<0000000014E099D0>"
tracemem(dt$Y)
#> [1] "<00000000135DEA38>"

# copy made of column X
dt2 <- dt[, .(X)]
#> tracemem[0x0000000014e099d0 -> 0x00000000174b0160]: copy [.data.table [ eval eval withVisible withCallingHandlers doTryCatch tryCatchOne tryCatchList tryCatch try handle timing_fn evaluate_call <Anonymous> evaluate in_dir block_exec call_block process_group.block process_group withCallingHandlers process_file <Anonymous> <Anonymous> <Anonymous> <Anonymous> do.call saveRDS withCallingHandlers

# no copy made
dt3 <- .subset(dt, "X")

# same pointer to SEXP
tracemem(dt3$X)
#> [1] "<0000000014E099D0>"
```

So the `.subset()` does not make a copy of the column, while the `data.table` selection does. You could paste the keys manually if you want, that's what I'm doing when you use `as.data.table = TRUE` in `read_fst()`, but that's not obvious to the average user.

If writing a subset of _rows_ efficiently has a use-case too, it might be interesting to add a `partial_write_fst()` method that can write subsets (rows and columns) without copying. That way you could split a data-set over a few `fst` files, just by selecting different _copy-windows_?

We could also add it to `write_fst` by adding two argument (e.g. `row_select` and `column_select`.

That might be interesting to have in a pipe as well:

```r
dt <- tibble(X = 1:26, Y = LETTERS) %>%
  slice(1:10) %>%
  select(Z = Y) %>%
  write_fst("1.fst")
```

these statements would require no copying at all if we could do row and column sub-setting in `fst`

thanks for sharing your ideas!

**Comment by mgahan** [comment]
Thanks everyone! This has been very educational. Appreciate all the hard work!

**Comment by MarcusKlik** [comment]
And thanks @mgahan for the feature request, I will post an update here once the feature is implemented!

**Comment by renkun-ken** [comment]
In my recent work, selecting both columns and rows are needed. 

I have a big `data.table` of nearly 100M rows and 1000+ columns. The data is a 1-sec snapshot of high frequency data, but I only need to store part of it (nearly 30M rows and 500 cols). Using a `dt[selector == TRUE, .(x1, x2, ...)]` would make a deep copy, costs a lot of time and memory even on a 80-core, 1TB RAM server, which is at least possible to do so. 

I will definitely appreciate it if fst accepts row/column selector vectors so that the data is stored without deep copy.

**Comment by MarcusKlik** [comment]
Hi @renkun-ken, thanks for your comments, you present a nice and important use-case for zero-copy writing, I'll add it to the `0.8.8` milestone!
